ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,249 @@
+      )
+  with Not_found -> debug "No valid token found in request"
+
+(* Slave Functions *)
+
+let send_notification (c:Context.t) reason p =
+  let record = (Eventgen.find_get_record p.tbl ~__context:c ~self:p.row) () in
+  begin match record with
+    | None -> debug "Could not send %s event for %s in %s table" reason p.row p.tbl;
+    | Some record -> Db_action_helper.events_notify ~snapshot:record p.tbl reason p.row;
+  end
+
+let make_change mtime tblname objref fldname newval =
+    Database.update (
+      (fun _ -> newval)
+      |> Row.update mtime fldname newval
+      |> Table.update mtime objref Row.empty
+      |> TableSet.update mtime tblname Table.empty)
+
+let make_empty_table mtime tblname =
+  Database.update (
+  (fun _ -> Table.empty)
+    |> TableSet.update mtime tblname Table.empty)
+
+let get_table_list db =
+  Db_cache_types.TableSet.fold_over_recent (-2L) (fun table stat value acc -> table::acc) (Db_cache_types.Database.tableset db) []
+
+let counter db table =
+  let t = Db_cache_types.TableSet.find table (Db_cache_types.Database.tableset db) in
+  {tblname=table; count=0; del=(Db_cache_types.Table.get_deleted_len t)}
+
+let object_count db =
+  let tables = get_table_list db in
+  List.rev (List.rev_map (counter db) tables)
+
+let count_eq_check (c1:count) (c2:count) =
+  c1.tblname = c2.tblname && c1.count = c2.count && c1.del = c2.del
+
+let count_check (c1_list: count list) (c2_list: count list) =
+  if List.length c1_list = List.length c2_list then
+    List.for_all2 count_eq_check c1_list c2_list
+  else
+    raise Content_length_required
+
+let flat_row_list update (rlist:row list) =
+  match rlist with
+  | [] -> [update]
+  | _ -> List.fold_left (fun acc (r:row) ->
+      (List.map (fun (f:field) -> {update with mtime=f.stat.modified; objref=r.objref; fldname=f.fldname; value=f.value;}) r.fields) @ acc) [] rlist
+
+let flatten_table_to_update (tbllist:table list) =
+  List.concat (List.map (fun (t:table) ->
+      (flat_row_list ({mtime=t.stat.modified; tblname=t.tblname; objref=""; fldname=""; value=Schema.Value.String "";})) t.rows) tbllist)
+
+let make_change_with_db_reply (update:update) =
+  match update.fldname with
+  | "" -> make_empty_table update.mtime update.tblname
+  | _ -> make_change update.mtime update.tblname update.objref update.fldname update.value
+
+let make_delete_with_db_reply __context delete db =
+  if Db_cache_impl.is_valid_ref (Context.database_of __context) delete.key then
+    Db_cache_types.remove_row delete.table delete.key db
+  else
+    begin
+      debug "Not valid ref";
+      db
+    end
+
+let apply_changes (delta:delta) __context =
+  if (Db_cache_types.Manifest.generation (Db_cache_types.Database.manifest !Xapi_slave_db.slave_db)) < delta.fresh_token then
+    begin
+      let new_db =
+        !Xapi_slave_db.slave_db
+        |> Database.set_generation delta.fresh_token
+        |> List.fold_right make_change_with_db_reply (flatten_table_to_update delta.tables)
```

Comment:
`List.fold_right` unfortunately isn't tail recursive (`List.fold_left` is). This would be a problem is the list can be long.

Reply:
Very true. I've changed this to use fold_left instead